core/src/queue.rs

```diff
@@ -205,53 +213,58 @@ impl Queue {
         &self,
         seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
         wsv: &WorldStateView,
+        expired_transactions: &mut Vec<VersionedAcceptedTransaction>,
     ) -> Option<VersionedAcceptedTransaction> {
-        loop {
-            let hash = self.signature_buffer.pop()?;
-            let entry = match self.txs.entry(hash) {
-                Entry::Occupied(entry) => entry,
-                // FIXME: Reachable under high load. Investigate, see if it's a problem.
-                Entry::Vacant(_) => continue,
-            };
-
-            match self.check_tx(entry.get(), wsv) {
-                Ok(MustUse(signature_check)) => {
-                    // Transactions are not removed from the queue until expired or committed
-                    seen.push(hash);
-                    if signature_check {
-                        return Some(entry.get().clone());
-                    }
-                }
-                Err(_) => {
-                    entry.remove_entry();
-                }
-            }
-        }
+        // NOTE: `SKIP_SIGNATURE_CHECK=false` because `signature_buffer` contains transaction which signature check can be either `true` or `false`.
+        self.pop_from::<false>(&self.signature_buffer, seen, wsv, expired_transactions)
     }
 
     /// Pop single transaction from the queue. Record all visited and not removed transactions in `seen`.
     fn pop_from_queue(
         &self,
         seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
         wsv: &WorldStateView,
+        expired_transactions: &mut Vec<VersionedAcceptedTransaction>,
+    ) -> Option<VersionedAcceptedTransaction> {
+        // NOTE: `SKIP_SIGNATURE_CHECK=true` because `queue` contains only transactions for which signature check is `true`.
+        self.pop_from::<true>(&self.queue, seen, wsv, expired_transactions)
+    }
+
+    /// Pop single transaction either from the queue or waiting buffer
+    #[inline]
+    fn pop_from<const SKIP_SIGNATURE_CHECK: bool>(
```

Comment:
Interesting move. This should well optimized by compiler, I believe

Reply:
I think so, at least i have tested this such approach and it worked ([link](https://godbolt.org/z/zc58d7GW5)).